### PR TITLE
fix(sandbox): acknowledge unchanged policy revisions

### LINKED ISSUE #2518: [Bug] Supervisor does not acknowledge newer sandbox policy revisions with an unchanged policy hash
## Agent Diagnostic

- Skills loaded: `create-github-issue` repository workflow
- OpenShell version tested: v0.0.91
- Latest release checked: v0.0.92; the settings poll loop still gates policy status on runtime hash or middleware changes
- Known fixes reviewed: #2159 / #2170 fixed initial policy acknowledgement, but not later same-hash revisions
- Possible duplicates reviewed: searched issues and pull requests for policy acknowledgement, same hash, pending revision, and supervisor polling terms; no matching report found
- Findings: a newer sandbox-scoped revision can be consumed for provider/config changes while remaining permanently unacknowledged when its effective policy hash is unchanged
- Remaining reason for filing: desired and current policy revisions can remain different even though the effective policy is already loaded

## Description

**Actual behavior:** after the supervisor has loaded sandbox policy revision 1 with hash `H`, settings polling can return sandbox policy revision 2 with the same hash `H`. This occurs when revision metadata or provider attachment state changes without changing effective policy content. The supervisor may refresh provider/config state, but `policy_runtime_changed` is false, so it never enqueues `PolicyStatusUpdate::loaded(2)`. The gateway continues to report desired revision 2 while current revision remains 1.

**Expected behavior:** a newer sandbox-scoped revision with the already-loaded effective policy hash should be acknowledged as loaded without rebuilding or reloading the identical OPA policy. If the revision also requires policy-runtime or middleware reconciliation, acknowledgement must wait for that reconciliation to succeed. Provider-environment refresh has its own revision/readiness lifecycle and should not incorrectly keep an already-loaded policy revision pending.

## Reproduction Steps

1. Start the supervisor with sandbox policy revision 1 and effective hash `H`.
2. Persist sandbox policy revision 2 with the same effective hash `H` and a newer provider/config revision.
3. Wake the settings poller (see related watcher issue #2517).
4. Observe that provider/config reconciliation runs.
5. Observe that no `LOADED` status is reported for policy revision 2 because `policy_runtime_changed` is false.
6. Query policy status and observe desired revision 2, current revision 1, and non-converged state.

A focused unit test can construct a `SettingsPollResult` with source `Sandbox`, version 2, and the current hash, then assert that revision 2 is selected for acknowledgement while equal/older revisions, different hashes, local overrides, and non-reloading origins are rejected.

## Environment

- OS: Linux arm64
- Runtime: Kubernetes sidecar topology
- OpenShell: v0.0.91
- Latest release checked: yes, v0.0.92
- Possible duplicates checked: yes

## Suspected Cause

The poll loop tracks `current_policy_hash` but not the version associated with that loaded hash. Its work gate and policy-status emission are driven by `policy_runtime_changed`, which intentionally remains false for identical hashes. Version convergence is therefore accidentally coupled to policy reload.

## Proposed Fix

Track the currently acknowledged policy version separately from the loaded policy hash. When a polling result:

- comes from `PolicySource::Sandbox`;
- has a version newer than the acknowledged version;
- has the same non-empty hash as the loaded runtime policy; and
- belongs to a policy origin that reloads gateway policy,

select it for acknowledgement without an OPA reload. Enqueue `PolicyStatusUpdate::loaded(version)` after any required policy-runtime or middleware reconciliation succeeds, update the tracked version, and emit the normal configuration-state success event. Keep provider-environment refresh and its readiness reporting independent from policy acknowledgement.

Do not apply this shortcut to local policy overrides, different hashes, equal/older revisions, or failed reconciliation.

## Logs

```text
loaded policy: version=1 hash=H
polled policy: version=2 hash=H source=Sandbox
provider environment: refreshed
reported loaded policy version: 1
```

### PR DESCRIPTION
**Pull request by NaveCohenMonday**
## Summary

Decouple sandbox policy-revision acknowledgement from OPA policy loading. A
newer sandbox revision whose effective policy hash is already loaded can now be
acknowledged without reloading identical policy content, allowing desired and
current revisions to converge.

## Related Issue

Fixes #2518

## Changes

- Track the acknowledged sandbox policy version separately from the loaded
  policy hash.
- Acknowledge only newer, sandbox-scoped revisions with the same non-empty
  effective policy hash.
- Wait for required policy-runtime or middleware reconciliation to succeed
  before acknowledging the revision.
- Exclude global policy, local overrides, equal or older versions, empty
  hashes, and different policy content from the shortcut.
- Emit acknowledgement success telemetry only after the gateway accepts the
  loaded-status report.
- Document the same-hash revision acknowledgement invariant.

## Testing

- [x] `CARGO_INCREMENTAL=0 mise run pre-commit` passes
- [x] Unit tests added/updated
- [ ] E2E tests added/updated (maintainer requested `test:e2e` after review)

Validation:

```text
CARGO_INCREMENTAL=0 mise exec -- cargo test -p openshell-sandbox --lib

test result: ok. 99 passed; 0 failed

CARGO_INCREMENTAL=0 mise exec -- cargo test -p openshell-sandbox \
  same_hash_ --lib

test result: ok. 3 passed; 0 failed
```

The scripted poll-loop tests cover `v1/H → v2/H` exactly-once acknowledgement
without an OPA reload, failed middleware reconciliation with no premature
status followed by retry and one acknowledgement, and the local/global,
changed/empty-hash, equal, and older-version scopes.

## Checklist

- [x] Follows [Conventional Commits](https://www.conventionalcommits.org/)
- [x] Commits are signed off (DCO)
- [x] Architecture docs updated

### CODE REVIEW

**Comment by NaveCohenMonday** [comment]
I have read the DCO document and I hereby sign the DCO.

File: crates/openshell-sandbox/src/lib.rs [hunk 1]
```diff
@@ -2931,6 +2968,27 @@ async fn run_policy_poll_loop(ctx: PolicyPollLoopContext) -> Result<()> {
...
         }
 
+        if let Some(version) = unchanged_policy_revision_ready_to_ack(
+            unchanged_policy_revision,
+            policy_runtime_changed,
+            policy_runtime_reconciled,
+        ) {
+            enqueue_policy_status(&status_sender, PolicyStatusUpdate::loaded(version));
+            current_policy_version = version;
+            ocsf_emit!(
```
Comment (drew) [inline crates/openshell-sandbox/src/lib.rs]: > **gator-agent**

**Warning:** This success OCSF event fires immediately after queueing the status update, before `report_policy_status` succeeds. A backlog or terminal RPC failure would therefore produce a false “Acknowledged” operational event. Carry the hash/event kind in `PolicyStatusUpdate` and emit this event from `run_policy_status_reporter` only after the RPC returns `Ok(())`; alternatively, record queueing here with ordinary debug tracing.

**Review by drew — COMMENTED** [review]
> **gator-agent**

## PR Review Status

Validation: This concentrated sandbox-policy fix directly addresses confirmed issue #2518 and has no duplicate candidate.
Head SHA: `182ef51768dfa66becb6847e0349170621cbddab`

Review findings:
- **Warning — missing stateful regression coverage:** the added predicate tests do not exercise the poll loop, so they cannot catch duplicate or premature acknowledgements. Please add a scripted-poll test proving `v1/H → v2/H` emits exactly one `LOADED(2)` without reloading OPA, a failed required middleware reconciliation emits nothing and retries, and a later success emits exactly one acknowledgement. Include the relevant negative scopes (local override, global source, changed/empty hash, and equal/older version).
- See the inline warning about emitting success telemetry before gateway acknowledgement succeeds.

Docs: The architecture invariant is documented. No Fern docs/navigation update is required because this fixes existing `--wait` semantics without changing a command or user workflow.

Tests: No local tests were run. The author reports pre-commit and two focused predicate tests passing; required CI is still waiting for the copy-PR mirror.

Next state: `gator:in-review` pending an author update. A `test:e2e` pipeline run will be required after review feedback is resolved because this changes sandbox policy-enforcement reconciliation.

File: crates/openshell-sandbox/src/lib.rs (hunk 1 above)
Comment (NaveCohenMonday) [reply crates/openshell-sandbox/src/lib.rs]: Addressed in c20b6b8c. PolicyStatusUpdate now carries an explicit success-event kind and policy hash. The reporter emits the acknowledgement OCSF event only after report_policy_status returns Ok(()); terminal delivery failures discard the update without success telemetry.

**Comment by NaveCohenMonday** [comment]
Addressed in `c20b6b8c`.

The poll loop now accepts a scripted gateway and middleware connector in tests,
while production continues to use the cached gRPC client and normal registry
connector. The new stateful coverage proves:

- `v1/H → v2/H` emits exactly one `LOADED(2)` and leaves the OPA generation
  unchanged;
- a failed required middleware reconciliation emits no `LOADED(2)`;
- the identical next poll retries reconciliation, and a later success emits
  exactly one `LOADED(2)`;
- local override, global source, empty hash, equal version, and older version
  do not use the shortcut;
- a changed hash follows the normal OPA reload and status path.

I also carried the acknowledgement event kind and policy hash in
`PolicyStatusUpdate`. The OCSF success event now emits from
`run_policy_status_reporter` only after `report_policy_status` returns
successfully.

Validation:

```text
CARGO_INCREMENTAL=0 mise exec -- cargo test -p openshell-sandbox --lib
test result: ok. 99 passed; 0 failed

CARGO_INCREMENTAL=0 mise run pre-commit
Finished successfully
```

I additionally mutation-checked the two critical guards: allowing equal
versions caused the exactly-once test to fail, and allowing acknowledgement
before runtime reconciliation caused the failed-middleware test to fail.

**Review by drew — COMMENTED** [review]
> **gator-agent**

## Re-check After Author Update

Thanks @NaveCohenMonday. I reviewed current head `c20b6b8c365dd5821594673d5fe06bf4cbe2ae93` after your update explaining the new scripted poll-loop coverage and post-RPC acknowledgement telemetry.

Validation: This concentrated sandbox-policy fix directly addresses confirmed issue #2518.
Head SHA: `c20b6b8c365dd5821594673d5fe06bf4cbe2ae93`

What I checked: the full current-head diff, the same-hash candidate and reconciliation gates, status-reporter delivery path, stateful positive/negative poll tests, and the architecture invariant.

Disposition: the two prior gator findings are resolved. The poll-loop tests now cover same-hash exactly-once acknowledgement, middleware failure/retry, and the requested negative scopes; success OCSF telemetry now follows a successful gateway status RPC. The independent review found no blocking correctness or security issue. It left one non-blocking determinism warning and one optional telemetry-test suggestion inline.

Docs: `architecture/sandbox.md` accurately documents the invariant. No Fern docs/navigation update is required because this fixes existing semantics without adding a direct UX workflow.

Tests: Code-only review; no local tests were run. A `test:e2e` pipeline run is required because this changes sandbox policy-enforcement reconciliation.

Next state: `gator:watch-pipeline`.

File: crates/openshell-sandbox/src/lib.rs [hunk 2]
```diff
@@ -3461,6 +3614,369 @@ filesystem_policy:
...
+            .await
+            .expect("policy report timed out")
+            .expect("policy reporter stopped");
+        assert_eq!(report, (version, true, String::new()));
+    }
+
+    async fn expect_no_policy_report(
+        reports: &mut tokio::sync::mpsc::UnboundedReceiver<(u32, bool, String)>,
+    ) {
+        assert!(
```
Comment (drew) [inline crates/openshell-sandbox/src/lib.rs:4108]: > **gator-agent**

**Warning (non-blocking):** `expect_no_policy_report` treats a 50 ms timeout as proof that the poll was processed without emitting a report. Under scheduler delay, the test can pass and abort before the scripted poll loop consumes the input. Add a notification when `poll_settings` dequeues each snapshot, await that barrier, then assert the report channel remains empty. This would make the exactly-once and negative-scope coverage deterministic.

File: crates/openshell-sandbox/src/lib.rs [hunk 3]
```diff
@@ -2422,18 +2508,31 @@ async fn run_policy_status_reporter(
             }
         }
 
-        if let Some(policy_hash) = update.initial_policy_hash {
+        if let Some(event) = update.success_event {
```
Comment (drew) [inline crates/openshell-sandbox/src/lib.rs:2636]: > **gator-agent**

**Suggestion:** The success event is correctly emitted only after `report_policy_status` returns `Ok`, but the scripted gateway always accepts reports. Consider a reporter-level test with queued failure/success outcomes and captured OCSF output to prove failures emit no acknowledgement success event and eventual success emits exactly one.

**Comment by NaveCohenMonday** [comment]
Rebased onto current `main` (`736e431d`) to resolve the conflict with the proxy-egress/runtime reconciliation changes from #2373.

The resolution preserves upstream's typed `PolicyValidation` vs `MiddlewareRegistry` failure handling and routes the existing injected test connector through that helper. It also handles the new overlap with fail-closed recovery: recovery takes precedence over the ordinary same-hash path, so restoring the last-known-good hash emits only one `LOADED` acknowledgement.

Validation on new head `4e6ed1ba77d3b45c62414a51328804e8cf917819`:

```text
cargo test -p openshell-sandbox
107 library tests, 5 binary tests, and 1 integration test passed

mise run pre-commit
passed
```

I also mutation-checked both conflict-sensitive guards: changing the newer-version comparison from `>` to `>=` fails the strict-scope test, and disabling recovery precedence fails its new assertion.

The force-push invalidated the previous copy-PR approval, so the new head will need `/ok to test 4e6ed1ba77d3b45c62414a51328804e8cf917819` from a maintainer.

**Review by johntmyers — COMMENTED** [review]
> **gator-agent**

## Re-check After Author Update

Thanks @NaveCohenMonday. I reviewed current head `4e6ed1ba77d3b45c62414a51328804e8cf917819` after your August 2 update explaining the rebase across the proxy-egress/runtime reconciliation changes and the fail-closed recovery overlap.

Head SHA: `4e6ed1ba77d3b45c62414a51328804e8cf917819`
Base SHA: `736e431d454c7de8a71e0fcdd3221ad6f9a552cb`
Merge base SHA: `736e431d454c7de8a71e0fcdd3221ad6f9a552cb`
Patch ID: `76105582c3c712f7f4d52b37605c7b6515c2d860`
Gator payload: `3`
Review mode: `follow_up`
Previous reviewed SHA: `c20b6b8c365dd5821594673d5fe06bf4cbe2ae93`

What I checked: the durable Gator feedback ledger, the author-only range-diff, the runtime/middleware conflict resolution, and recovery precedence for a restored last-known-good hash. Drew's prior findings and their resolutions remain preserved; I did not reopen or duplicate them.

Disposition: resolved. The rebase preserves the prior acknowledgement and post-RPC telemetry fixes, recovery takes precedence over the ordinary same-hash acknowledgement path, and the independent follow-up review found no new blocking findings.

Blocking findings:
- No blocking findings remain.

Carried findings:
- None. The prior Gator findings remain resolved.

Docs: `architecture/sandbox.md` remains sufficient; this fixes existing behavior without introducing a new direct UX workflow.

Tests: `test:e2e` remains required because this changes sandbox policy-enforcement reconciliation. I posted `/ok to test 4e6ed1ba77d3b45c62414a51328804e8cf917819` for the rebased head so the copy-PR pipeline can proceed.

Next state: `gator:watch-pipeline`.

**Comment by johntmyers** [comment]
@NaveCohenMonday you're hitting some flakey tests which have been turned off as of today. i'd rebase to see if we can get the pipeline green so we can prep to merge

**Comment by johntmyers** [comment]
> **gator-agent**

## Author Follow-Up Nudge

This PR has been in `gator:in-review` for more than 48 business hours waiting on the requested rebase after the flaky branch-check failure.

@NaveCohenMonday, please rebase onto current `main` and push the updated head so the branch pipeline can rerun. If this is no longer planned, please say so and a maintainer can close it out.

**Comment by NaveCohenMonday** [comment]
Rebased onto current main (4cb77a90) as requested and force-updated the existing fork branch.

New head: 25202b2671f2acafae824e699f21f660210e179b

The two-commit range-diff is patch-identical to the previously reviewed head, and both commits retain the Nave Cohen DCO sign-off.

Validation on the rebased head:

- CARGO_INCREMENTAL=0 cargo test -p openshell-sandbox --lib: 107 passed, 0 failed
- CARGO_INCREMENTAL=0 mise run pre-commit: passed
- git diff --check: passed

This fork head now needs /ok to test 25202b2671f2acafae824e699f21f660210e179b so the copy-PR branch checks can rerun.

**Comment by johntmyers** [comment]
> **gator-agent**

## Re-check After Author Update

Thanks @NaveCohenMonday. I checked current head `25202b2671f2acafae824e699f21f660210e179b` after your August 9 update explaining the requested rebase and patch-equivalent range-diff.

Head SHA: `25202b2671f2acafae824e699f21f660210e179b`
Base SHA: `4cb77a900ebd6b789d2b68daaba4830866833b1c`
Merge base SHA: `4cb77a900ebd6b789d2b68daaba4830866833b1c`
Patch ID: `76105582c3c712f7f4d52b37605c7b6515c2d860`
Gator payload: `3`

What I checked: the durable Gator feedback ledger, current patch identity, prior marked reviews and inline findings, and the current required-check state. The ledger confirms this rebase is patch-equivalent to the last reviewed head, so I did not re-review the unchanged effective patch. Drew's original findings and their later resolutions remain preserved.

Disposition: resolved. No blocking review findings remain.

Tests: `test:e2e` remains required because this changes sandbox policy-enforcement reconciliation. I posted `/ok to test 25202b2671f2acafae824e699f21f660210e179b` so the copy-PR pipeline can proceed.

Next state: `gator:watch-pipeline`.

**Review by johntmyers — APPROVED** [review]

**Comment by johntmyers** [comment]
> **gator-agent**

## Monitoring Complete

Monitoring is complete because this PR has merged.

Head SHA: `25202b2671f2acafae824e699f21f660210e179b`
Gator payload: `3`

Final status: Gator preserved Drew's original review and inline findings, their subsequent resolutions, and the patch-equivalent same-SHA disposition. Required branch, Helm, and E2E gates passed, maintainer approval was present, and the PR merged from `gator:merge-ready`.

I removed the active `gator:*` label because there is nothing left for gator to monitor on this PR.
